cluster/write-file.go

```diff
@@ -0,0 +1,68 @@
+package cluster
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"github.com/docker/docker/api/types/container"
+	"github.com/rancher/rke/docker"
+	"github.com/rancher/rke/hosts"
+	"github.com/rancher/rke/log"
+	"github.com/sirupsen/logrus"
+	"path/filepath"
+)
+
+// WriteFile - Writes file to Node.
+func (c *Cluster) WriteFile(ctx context.Context, host *hosts.Host, role string, path string, mode string, content string) error {
+	log.Infof(ctx, "[%s] Deploying file %s on node [%s]", role, path, host.Address)
+	err := c.doWriteFile(ctx, host, role, path, mode, content)
+	if err != nil {
+		return fmt.Errorf("[%s] Failed to file %s on node [%s]: %v", role, path, host.Address, err)
+	}
+	logrus.Debugf("[%s] Successfully deployed %s on node [%s]", role, path, host.Address)
+
+	return nil
+}
+
+func (c *Cluster) doWriteFile(ctx context.Context, host *hosts.Host, role string, path string, mode string, content string) error {
+	isRunning, err := docker.IsContainerRunning(ctx, host.DClient, host.Address, "deploy-file", true)
+	if err != nil {
+		return err
+	}
+	if isRunning {
+		err = docker.DoRemoveContainer(ctx, host.DClient, "write-file", host.Address)
+		if err != nil {
+			return err
+		}
+	}
+
+	b64Content := base64.StdEncoding.EncodeToString([]byte(content))
+	pathDir := filepath.Dir(path)
+	imageCfg := &container.Config{
+		Image: c.SystemImages.Alpine,
+		Cmd: []string{
+			"sh",
+			"-c",
+			fmt.Sprintf("cp -f %s %s-$(date -Iseconds -u); echo \"%s\" | base64 -d > %s && chmod %s %s", path, path, b64Content, path, mode, path),
```

Comment:
We have been thinking about using docker cp API to write files to nodes. It should be faster, much easier to report errors, and you don't need to start() the container, you just need to create() it. Do you think it would make sense to start using that here?

Reply:
That's an interesting idea. I think the one advantage to this method is that I'm creating a backup of the previous file before writing over it. I'm not sure you can could you do that with docker cp.